### DISCUSSION TITLE
Does Next.js support pm2 cluster mode?

### DISCUSSION THREAD

**Discussion by choisium** [vercel/next.js#10675] [Help] [answered]
Hi guys,
I'm trying to use next.js with pm2 cluster mode to deploy, but it always returns error.
So I made a simple express app, but it works well with pm2 cluster mode!
I wonder what's happening.. Is it possible to use pm2 cluster mode with next.js? Or there is something I missed?

My pm2 version is 4.2.3, and express version is 4.17.1. next and related version is below.
Below is my code and some screenshots of my error.

```javascript
// pages/index.js
export default function Index() {
    return (<div><p>Hello Next.js</p></div>)
}
```

```json
// package.json
{
  "name": "test",
  "version": "1.0.0",
  "description": "",
  "dependencies": {
    "next": "^9.2.2",
    "react": "^16.12.0",
    "react-dom": "^16.12.0"
  },
  "devDependencies": {},
  "scripts": {
    "dev": "next",
    "build": "next build",
    "start": "next start"
  },
  "author": "",
  "license": "ISC"
}
```

when I run the next server
![next](https://user-images.githubusercontent.com/35210772/75206242-5089dd00-57b9-11ea-9269-0c9a2de2f190.png)
I found there is no error in id 0. All error occured in id 1.
Also it works well when I make only one instance..
I thought the problem is related to the port, but isn't pm2 handle this?

```javascript
// test.js (express server)
const express = require('express')
const app = express()
const uuid = require('uuid')
var instance_id = uuid.v4()
const port = 3000
app.get('/', function (req, res) { 
  res.send(`Hello World! ${instance_id}`)
})
app.listen(port, function () {
  console.log(`application is listening on port ${port}...`)
})
```

when I run the express server
![express](https://user-images.githubusercontent.com/35210772/75206247-541d6400-57b9-11ea-998d-d5e74fdc4611.png)

I posted this on [spectrum](https://spectrum.chat/next-js/general/next-js-9-pm2-cluster-mode~4dad85e3-1089-4171-b021-1cbb439dd0f6), too.
Also let me know if this should be in Issues. I'm not sure where it should be..

**Comment by vhadjiev** [comment]
I had similar problem (not with pm2, but the error was the same) and the issue in my case was that the version of Node was outdated. Can check what version of Node are you using and update to more recent one?

**Reply by choisium** [reply]
![image](https://user-images.githubusercontent.com/35210772/75230455-fad42580-57f6-11ea-9819-44c20160db7b.png)
Well I'm using node version 12.14.0. Is this outdated..?

**Reply by vhadjiev** [reply]
Latest stable is 12.16.1 but your version is relatively up-to-date so maybe the problem you have is different than my.

**Reply by choisium** [reply]
Alright.. thank you!

**Comment by timneutkens** [comment]
Did you check the logs outlined there, it shows only 15 lines which is not enough to know why the command failed to boot. Looks like it should be located here: `/Users/choisium/.pm2/logs/next-error-1.log`

**Reply by choisium** [reply]
![image](https://user-images.githubusercontent.com/35210772/75230396-e5f79200-57f6-11ea-8d13-8551afeeaee1.png)
There was no other error message..

**Comment by timneutkens** [comment, marked as answer]
1. Install `pm2` in the dependencies
2. Make sure you have a package.json `scripts` section like this:

```json
"scripts": {
  "dev": "next",
  "build": "next build",
  "start": "next start",
  "pm2": "pm2"
}
```

3. Add `pm2.json` and add the following configuration:

```json
{
  "apps": [
      {
          "name": "my-app",
          "script": "npm",
          "args" : "start",
          "instances": "max",
          "exec_mode": "cluster"
      }
  ]
}
```

4. Make sure you run `npm run build` to build the Next.js app for production
5. Run `npm run pm2 start pm2.json`


> Note: `pm2.json` is doesn't have to have that exact name, it's just what I used while testing.

The configuration above will start as many instances as possible and far as I tried works correctly.

---

I would highly recommend looking into hosting on https://zeit.co/ though it's automatically optimized for Next.js in many ways: https://nextjs.org/docs/deployment#optimized-for-nextjs

**Reply by choisium** [reply]
I followed exactly as you told me except the number of instances(just for test I changed it to 2), but the result was the same.. Anyway thank you for your help!
![image](https://user-images.githubusercontent.com/35210772/75303286-a32fcb80-5883-11ea-8882-82261963aff7.png)

**Reply by timneutkens** [reply]
Make sure you run `npm run build` before bootup

**Reply by choisium** [reply]
I did run build before start pm2, but the same error occured again.

**Reply by timneutkens** [reply]
I ran the steps with a plain Next.js application and it worked fine so I'm really not sure what's wrong in your case.

**Reply by choisium** [reply]
Then at least I can be sure that the problem is on my side, not next.js! Thank you so much!!

**Reply by vlatkoodrljin** [reply]
@choisium did you find what was your source of problem?

@timneutkens I can confirm you that I have the same issue with all your steps applied. Latest next, Node v12.12.0, NPM 6.11.3

Please make more clear in docs how to deploy scalable nextjs app using multi-core cpu architecture (cluster mode with process management tool). Without that it's hard to reliably setup anything meaningful.
It's frustrating to have such a great solution for building isomorphic apps like nextjs and to struggle with proper/more flexible deploying.

Tnx, and I hope that we will make this situation clear.

**Reply by alexandrebodin** [reply]
Hi, here is the only way I could make this work without having issues:

`ecosystem.config.js`
```js
module.exports = {
  apps: [{
    //...
    script: './node_modules/.bin/next',
    args: 'start -p my_port'
  }]
};
```

This is the only way to get next.js logs in the pm2 logs and see any errors. this was also the only way I found to make sure port bindings worked correctly. using npm scripts make pm2 fail at gracefull reloads

**Reply by ja-klaudiusz** [reply]
@timneutkens I ran your steps with a WITH-REDUX-SAGA-APP example and it doesn't work. 
![image](https://user-images.githubusercontent.com/57961329/89284747-87f59300-d64f-11ea-86fc-d48d08af9dff.png)
I run this on windows, what can cause this error?

**Reply by kvedantmahajan** [reply]
If something from @timneutkens reply is not working maybe do `pm2 kill` and then test. This one consumes less memory too as compared to @alexandrebodin `ecosystem.config.js` script.

**Reply by constsz** [reply]
Hey @timneutkens. thanks so much, worked perfectly!

**Reply by esmalleydev** [reply]
> Hi, here is the only way I could make this work without having issues:
> 
> `ecosystem.config.js`
> 
> ```js
> module.exports = {
>   apps: [{
>     //...
>     script: './node_modules/.bin/next',
>     args: 'start -p my_port'
>   }]
> };
> ```
> 
> This is the only way to get next.js logs in the pm2 logs and see any errors. this was also the only way I found to make sure port bindings worked correctly. using npm scripts make pm2 fail at gracefull reloads

This worked for me as well... but does anyone know what the root issue here is?

**Reply by quyphan97** [reply]
@esmalleydev please help me
<img width="599" height="306" alt="image" src="https://github.com/user-attachments/assets/d1dffa90-9367-4bc2-b4b5-650ed3acd51f" />

Do you have any other solution? I updated it like that and it worked but the memory increased 56mb -> >500mb and continues to increase
<img width="971" height="205" alt="image" src="https://github.com/user-attachments/assets/1b323594-2fca-4ec1-8145-402b060622fc" />

**Reply by esmalleydev** [reply]
I'd suggest just writing your own nextjs server using the default node cluster mode

**Reply by quyphan97** [reply]
Yes, thank you very much!

**Comment by magnnus** [comment]
```js
"scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "pm2": "pm2-runtime start pm2.config.js --env production"
  }
```

when i run `npm run pm2`, I got an **error**:

```
error - Failed to load env from .env Error: ENOTDIR: not a directory, stat '/Users/my-project-dir/pm2.config.js/.env'
```

but, if I replace pm2 script to `pm2 start pm2.config.js --env production`,  I succeed.

The pm2.config.js is:

```js
module.exports = {
  apps : [
    {
      name: 'next',
      script: 'npm',
      args: 'start',
      instances: 4,
      exec_mode: 'cluster'
    }
  ]
};
```

**Reply by ong-ar** [reply]
my solution.

```
      script: 'node_modules/next/dist/bin/next',
      args: 'start',
      cwd: './',
```

**Reply by michaelschufi** [reply]
~I can confirm this is working with a `pm2.json` with the content:~

```
{
  "apps": [
    {
      "name": "my-app",
      "script": "node_modules/next/dist/bin/next",
      "args": "start",
      "cwd": "./",
      "instances": "max",
      "exec_mode": "cluster"
    }
  ]
}
```

and then 
`pm2-runtime start pm2.json`

Edit: Somehow `pm2-runtime` passes the wrong working directory. Running `pm2` before `pm2-runtime` solved it for me.

```
pm2 start pm2.json
pm2-runtime start pm2.json
```

**Reply by magnnus** [reply]
@michaelschufi Thx. I have tried your solution, It's OK.   
And I find that, if I first run `pm2 list` before start the app(`pm2-runtime start pm2.json`), It's OK too.

**Reply by seungdols** [reply]
When will this issue be fixed? Same for next 11.0.1 version.

@magnnus Do you know why I need to run pm2 list first?

**Reply by chalermpong** [reply]
@seungdols I just found this issue. This might be a problem in pm2-runtime when no daemon is running. See https://github.com/Unitech/pm2/issues/4950.

**Comment by asheroto** [comment]
I had nothing but trouble trying to use `pm2.config.js` and only had luck using `pm.json`

I did not have a problem where I had to run `pm2` before anything else.

This is what is working for me:

`package.json` scripts block:
```
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "export": "next export",
    "pm2": "pm2 start pm2.json",
  },
```

`pm2.json`:
```
{
	"apps": [
		{
			"name": "app-name",
			"script": "node_modules/next/dist/bin/next",
			"args": "start",
			"cwd": "./",
			"instances": "max",
			"exec_mode": "cluster"
		}
	]
}
```

I can now run `npm run pm2` from the command line, or click on the play button next to `pm2` in VS Code.  Both work.

Success.

**Comment by marnixhoh** [comment]
When Next.js is run in a cluster, does every process then have its own copy of things such as statically generated files, optimized images (next/image), etc.? Or will every process share these?

**Reply by asheroto** [reply]
pm2 cluster mode shares the files :-)

**Reply by marnixhoh** [reply]
Oh that's awesome! Thanks for your quick reply! :D

**Comment by smasilamani-cfins** [comment]
Does anyone know how we can run it in cluster mode inside docker? I have tried all the steps mentioned here but nothing works inside docker container.

My Docker command looks like this : CMD [ "pm2-runtime" ,"start","pm2.json"]

pm2.json file

```
{
  "apps": [
    {
      "name": "App-Client",
      "cwd": "/usr/app/client",
      "script": "node_modules/next/dist/bin/next",
      "args": "start -p 4000",
      "instances": "max",
      "exec_mode": "cluster",
      "kill_timeout": 5000,
      "env": { "NODE_ENV": "production", "production": true }
    }
  ]
}
```

Error: (As you can see its looking for .next folder inside pm2.json folder)

> 2021-12-03T04:31:51: PM2 log: App [App-Client:0] exited with code [1] via signal [SIGINT]
> 2021-12-03T04:31:51: PM2 log: App [App-Client:0] starting in -cluster mode-
> Error: Could not find a valid build in the '/usr/app/client/pm2.json/.next' directory! Try building your app with 'next build' before starting the server.
>     at Server.readBuildId (/usr/app/dip/client/node_modules/next/next-server/server/next-server.ts:1671:15)
>     at new Server (/usr/app/dip/client/node_modules/next/next-server/server/next-server.ts:179:25)
>     at createServer (/usr/app/dip/client/node_modules/next/server/next.ts:41:10)
>     at start (/usr/app/dip/client/node_modules/next/server/lib/start-server.ts:9:15)
>     at exec (/usr/app/dip/client/node_modules/next/cli/next-start.ts:53:3)
>     at /usr/app/dip/client/node_modules/next/bin/next.ts:107:36
> 2021-12-03T04:31:51: PM2 log: App name:App-Client id:1 disconnected
>

**Reply by hugw** [reply]
@saachinsiva not sure if you are still facing the issue with Next.JS integrated with Docker and PM2...

The problem is only `pm2-runtime`.
In order for you to overcome the issue, try running `pm2 start` first.

Here is what I did...

Dockerfile:
```
...
CMD ["npm", "run", "pm2"]
```

package.json:
```
...
"scripts: {
  ...
  "pm2": "pm2 start pm2.config.js && pm2-runtime pm2.config.js",
}
```

It's important that pm2 is executed before pm2-runtime.

Then my pm2.config.js:
```
module.exports = {
  apps: [
    // Main Application
    {
      name: 'my-app',
      script: './node_modules/next/dist/bin/next',
      args: 'start',
      exec_mode: 'cluster',
    },
  ],
}
```

**Reply by smasilamani-cfins** [reply]
Thank you. I will try this and let you know.

**Reply by EmitoSanti** [reply]
It work in my instance ec2 AWS with docker + pm2 + nextjs 14.2.5
 Docker build and ducker run success

ecosystem.config.js
```
module.exports = {
  apps: [
    {
      name: "intelifaz-web",
      script: "./node_modules/next/dist/bin/next",
      args: 'start',
      node_args : '-p 3033',
      exec_mode: "cluster",
      watch: false,
      instances: 4
... env...
    },
  ],
};
```


DockerFile
```
`FROM node:18.19.0-slim
LABEL version="1.0"
LABEL description="This is the base docker image for prod frontend next app. Repo: intelifaz-web"

WORKDIR /app

#Install dependencies
COPY package*.json ./

RUN npm install
# Bundle app source
COPY . .

# Construye la aplicación Next.js
# RUN npm run build:prod

# Instala PM2 globalmente
RUN npm install pm2 -g

// # Port
EXPOSE 3033

# this run when the image is run in the container
# CMD [ "pm2-runtime", "start", "npm", "--", "start", "--name", "intelifaz-web", "--instances", "4", "--exec_mode", "cluster" ]
# CMD [ "pm2-runtime", "ecosystem.config.json" ]
# CMD ["sh", "-c", "pm2-runtime --instances 4 start npm start"]
# CMD ["pm2-runtime", "start", "npm", "--instances", "4", "--" , "start"]

# CMD ["pm2-runtime", "start", "npm"]

# CMD ["/bin/sh", "-c", "pm2-runtime start 'npm start' -i 2"]

# CMD ["/bin/sh", "-c", "pm2-runtime start ecosystem.config.json"]

CMD ["npm", "run", "pm2"]`
```

```
package.json
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "build:dev": "env-cmd -f .env.development next build",
    "build:prod": "env-cmd -f .env.production next build",
    "start": "next start",
    "lint": "next lint",
    "pm2": "pm2 start ecosystem.config.js && pm2-runtime ecosystem.config.js"
  },

```

Thank for this line -> "pm2": "pm2 start ecosystem.config.json && pm2-runtime ecosystem.config.json"
you saved my weekend

**Comment by saaiful** [comment]
![image](https://user-images.githubusercontent.com/4525049/181877955-60265e7b-f61c-42ed-8d5c-229bd1585ca2.png)
pm2 in cluster mode keeps only one instance running, other instance keeps stopping with error! What to do?

**Reply by stegano** [reply]
In my case it works fine on pm2 when standalone.

Try this,

Step 1. Enable the value `experimental.outputStandalone` in your `next.config.js` file.
```js
// next.config.js
module.exports = {
  ...
  experimental: {
    outputStandalone: true,
  }
}
```

Step 2. Build your app
```bash
next build
```

Step 3. Run Nextjs app (run standalone mode) with pm2
```bash
pm2-runtime start ./node_modules/.bin/next -i 2
```

**Reply by saaiful** [reply]
Still same, all process keeps crashing.

**Reply by DuckThom** [reply]
I know it's an old comment but to anyone else coming across this. The processes are running in `fork` mode, not `cluster` mode. If they were originally started as a fork, you need to `pm2 delete` them first before running `pm2 start`. The start command does not update the mode of already running processes when the ecosystem file is updated.

When running in `fork` mode, all the processes try to use port 3000 and only 1 of them is able to do so. Which is why there's only 1 running process in the screenshot above.

**Comment by saaiful** [comment]
I solved this issue a while ago.  My `ecosystem.config.js` file looks like this:
```js
module.exports = {
    apps: [
      {
        name: 'app-name',
        script: "node_modules/next/dist/bin/next",
        args: "start",
        instances: 4, 
        exec_mode: 'cluster',
        autorestart: true,
      },
    ],
};
```
Afterward, I deleted all PM2 processes using `pm2 delete`  and then started PM2 with `pm2 start` without passing any arguments. It's now working without any issues.

**Reply by swarm-agent** [reply]
> I solved this issue a while ago. My `ecosystem.config.js` file looks like this:
> 
> ```js
> module.exports = {
>     apps: [
>       {
>         name: 'app-name',
>         script: "node_modules/next/dist/bin/next",
>         args: "start",
>         instances: 4, 
>         exec_mode: 'cluster',
>         autorestart: true,
>       },
>     ],
> };
> ```
> 
> Afterward, I deleted all PM2 processes using `pm2 delete` and then started PM2 with `pm2 start` without passing any arguments. It's now working without any issues.

Let's fucking go baby. Good job bro

**Comment by superprat** [comment]
Is anyone facing an issue getting console.log to show up in pm2 logs with this method?

Perhaps using a custom express server?
